nym-vpn-app/src-tauri/src/grpc/tunnel.rs

```diff
@@ -0,0 +1,264 @@
+use serde::Serialize;
+use time::OffsetDateTime;
+use ts_rs::TS;
+
+use nym_vpn_proto as p;
+use p::tunnel_state::{ActionAfterDisconnect, ErrorStateReason, State};
+
+#[derive(Serialize, Clone, Debug, PartialEq, TS)]
+#[ts(export)]
+#[serde(rename_all = "camelCase")]
+pub struct WgNode {
+    pub endpoint: String,
+    pub public_key: String,
+    pub private_ipv4: String,
+    pub private_ipv6: String,
+}
+
+#[derive(Serialize, Clone, Debug, PartialEq, TS)]
+#[ts(export)]
+#[serde(rename_all = "camelCase")]
+pub struct MixnetData {
+    pub nym_address: Option<String>,
+    pub exit_ipr: Option<String>,
+    pub ipv4: String,
+    pub ipv6: String,
+}
+
+#[derive(Serialize, Clone, Debug, PartialEq, TS)]
+#[ts(export)]
+#[serde(rename_all = "camelCase")]
+pub struct WireguardData {
+    pub entry: WgNode,
+    pub exit: WgNode,
+}
+
+#[derive(Serialize, Clone, Debug, PartialEq, TS)]
+#[ts(export)]
+#[serde(untagged)]
+pub enum TunnelData {
+    Mixnet(MixnetData),
+    Wireguard(WireguardData),
+}
+
+#[derive(Serialize, Clone, PartialEq, Debug, TS)]
+#[ts(export)]
+#[serde(rename_all = "camelCase")]
+pub struct Tunnel {
+    pub entry_gw_id: String,
+    pub exit_gw_id: String,
+    pub connected_at: Option<i64>,
+    pub data: TunnelData,
+}
+
+#[derive(Default, Debug, Clone, Serialize, PartialEq, TS, strum::Display)]
+#[ts(export)]
+#[serde(rename_all = "camelCase")]
+pub enum TunnelState {
+    #[default]
+    Disconnected,
+    Connected(Tunnel),
+    Connecting(Option<Tunnel>),
+    Disconnecting(Option<TunnelAction>),
+    Error(TunnelError),
+    Offline {
+        reconnect: bool,
+    },
+}
+
+impl TunnelState {
+    pub fn from_proto(tunnel: &State) -> Result<TunnelState, &str> {
+        Ok(match tunnel {
+            State::Disconnected(_empty) => TunnelState::Disconnected,
+            State::Connecting(c) => {
+                let tunnel = c
+                    .connection_data
+                    .as_ref()
+                    .map(Tunnel::try_from)
+                    .transpose()?;
+                TunnelState::Connecting(tunnel)
+            }
+            State::Connected(c) => {
+                let tunnel = c
+                    .connection_data
+                    .as_ref()
+                    .map(Tunnel::try_from)
+                    .transpose()?
+                    .ok_or("missing tunnel data")?;
+                TunnelState::Connected(tunnel)
+            }
+            State::Disconnecting(action) => {
+                TunnelState::Disconnecting(TunnelAction::from_proto(action.after_disconnect()))
+            }
+            State::Error(e) => TunnelState::Error(e.reason().into()),
+            State::Offline(o) => TunnelState::Offline {
+                reconnect: o.reconnect,
+            },
+        })
+    }
+}
+
+impl From<&p::WireguardNode> for WgNode {
+    fn from(p_data: &p::WireguardNode) -> Self {
+        WgNode {
+            endpoint: p_data.endpoint.clone(),
```

Comment:
Implementing `From<p::WireguardNode>` instead would be nicer and would be more straightforward for call site that the data is not borrowed but actually cloned.